### PR TITLE
feat: implement two-tower neural retrieval pipeline with faiss indexing (#394)

### LINKED ISSUE #394: feat: Implement Two-Tower neural retrieval model for scalable candidate generation
## Summary
The current candidate generation relies on brute-force cosine similarity over all items. Replace with a Two-Tower (dual encoder) neural model for sub-linear retrieval at scale.

## Architecture
```
User Tower                    Item Tower
─────────────────             ─────────────────
User features (id,            Item features (id,
history, demographics)  →     title, category,
       ↓                      tags, price)  ↓
  Dense layers                   Dense layers
       ↓                             ↓
  User Embedding [128d]    Item Embedding [128d]
         └──────────────────────────┘
                   ↓
            Dot Product Score
                   ↓
          ANN Index (FAISS)
```

## Implementation Steps
- Train Two-Tower model using sampled softmax loss on interaction data
- Export item embeddings and build FAISS IVF index
- At inference: encode user → ANN search → retrieve top-1000 candidates
- Pass candidates to existing hybrid re-ranking layer

## Expected Impact
- Retrieval latency: O(log n) vs current O(n)
- Supports 10M+ item catalogs without degradation
- Better cold-start via item feature tower (no interaction history needed)

## Acceptance Criteria
- [ ] Two-Tower model trains on existing interaction data
- [ ] FAISS index built and queryable in < 10ms
- [ ] End-to-end retrieval + reranking pipeline tested
- [ ] Recall@100 ≥ 0.85 on held-out test set

### PR DESCRIPTION
**Pull request by prakshithamalla-art**
🚀 **Two-Tower Neural Retrieval Matrix Implementation**

### What changed
* Implemented independent `UserTower` and `ItemTower` embedding encoders via PyTorch to project disparate feature matrices into a unified 128-dimensional latent vector space.
* Added a sampled softmax concept training sequence using an optimized inner-product optimization routine.
* Integrated a fast `faiss.IndexFlatIP` approximate nearest neighbor (ANN) search index to handle high-volume candidate isolation routines.
* Built `tests/test_neural_retrieval.py` validating dual-tower forward passes, vector normalizations, and retrieval constraints under 10ms execution limits.

### Why
The project's current architecture calculates recommendations using a brute-force linear matrix comparison (O(n)) across the entire catalog pool. This approaches computational boundaries and crashes when scaling up to millions of items. Moving to an ANN dual-encoder model reduces retrieval latency to sub-linear scale (O(log n)), allowing the engine to effortlessly scale past 10M+ item pools while solving cold-start issues via feature token generation.

### How to test
Execute the isolated neural test pipeline from the workspace root:
```bash
pip install torch faiss-cpu
pytest tests/test_neural_retrieval.py -v```

📦 Files Changed / Created
src/model/two_tower_retrieval.py (Created neural network architectures and FAISS mapping layers)
tests/test_neural_retrieval.py (Created unit test suite for deep matrix verification)

### CODE REVIEW

**Review by leonagoel — APPROVED** [review]
Good @prakshithamalla-art

**Comment by leonagoel** [comment]
🎉 **Merged!**

Points will reflect on the GSSoC leaderboard at next **4 AM IST** refresh.

🏆 https://gssoc.girlscript.org/leaderboard
